### PR TITLE
feat: add datasource for AWS Lambda Layers

### PR DESCRIPTION
**Pull request by kayman-mk**
## Changes

This PR adds a new datasource to update the versioned layer arns for AWS Lambda Layers.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/aws-versioned-arn/readme.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+This datasource returns releases from a versioned AWS ARN. It currently supports Lambda Layer ARNs only.
```
Comment (HonkingGoose) [inline lib/modules/datasource/aws-versioned-arn/readme.md:1]:
```suggestion
This datasource returns releases from a versioned AWS ARN.
It currently supports Lambda Layer ARNs only.
```

Our docs follow the "one sentence per line pattern". 😉

**Comment by kayman-mk** [comment]
@HonkingGoose Thanks for your review. I committed your change.

I still have two questions which I can't answer. How does Renovate know what to put into the `serializedArnFilter` of my datasource (how to get the `runtime`, `architecture` and `depName` from the code into my function)? And how does Renovate know, which part of my string has to be updated? I simply do not understand how it works.

Could you please clarify?

```hcl
locals {
  # renovate: datasource=aws-versioned-arn architecture=x86_64 runtime=python37
  insight_layer_arn = "arn:aws:lambda:us-east-1:580247275435:layer:LambdaInsightsExtension:21"
}

resource "aws_lambda_function" "example" {
  # ... other configuration ...

  layers = [local.insight_layer_arn]
}
```


```json
{
  "regexManagers": [
    {
      "fileMatch": ["\\.tf$"],
      "matchStrings": [
        ".*renovate: datasource=aws-versioned-arn architecture=(?<architecture>.*) runtime=(?<runtime>.*)\\s+.* = \"(?<depName>.*):(?<currentValue>\\d+)\""
      ],
      "versioningTemplate": "loose"
    }
  ]
}
```

```js
 async getReleases({
    packageName: serializedArnFilter,
  }: GetReleasesConfig): Promise<ReleaseResult | null> {
    const lambdaLayerVersions = await this.getSortedLambdaLayerVersions(
      serializedArnFilter
    );
}
```

**Comment by HonkingGoose** [comment]
I'll let the maintainers answer your questions. 😉

**Comment by kayman-mk** [comment]
@JamieMagee Could you step in please?

File: lib/modules/datasource/aws-versioned-arn/index.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      LayerName: JSON.parse(serializedLambdaLayerFilter),
+    });
+
+    const matchingLayerVersions = await this.lambda.getValue().send(cmd);
+    matchingLayerVersions.LayerVersions =
+      matchingLayerVersions.LayerVersions ?? [];
+
+    return matchingLayerVersions.LayerVersions.sort((layer1, layer2) => {
+      return (layer1.Version ?? 0) - (layer2.Version ?? 0);
+    });
```
Comment (viceice) [inline lib/modules/datasource/aws-versioned-arn/index.ts:41]: I don't think we need to sort here, the upper layer will sort by version anyways.

File: lib/modules/datasource/aws-versioned-arn/readme.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+Here's an example of using the regex manager to configure this datasource:
+
+```json
+{
+  "regexManagers": [
+    {
+      "fileMatch": ["\\.tf$"],
+      "matchStrings": [
+        ".*renovate: datasource=aws-versioned-arn architecture=(?<architecture>.*) runtime=(?<runtime>.*)\\s+.* = \"(?<depName>.*):(?<currentValue>\\d+)\""
```
Comment (viceice) [inline lib/modules/datasource/aws-versioned-arn/readme.md:41]: Custom regex groups are not saved. please checkout `aws-machine-image` for how to encode those in `packageName`.

Comment (kayman-mk) [reply lib/modules/datasource/aws-versioned-arn/readme.md:41]: Thanks, viceice. I've seen that. Thought that there might be another way to pass the information.

Comment (viceice) [reply lib/modules/datasource/aws-versioned-arn/readme.md:41]: no other way. 🙃

File: lib/modules/datasource/aws-versioned-arn/index.ts (hunk 2 above)
Comment (kayman-mk) [reply lib/modules/datasource/aws-versioned-arn/index.ts]: The AWS API does not say anything about the sort order. So I think it is better to sort. Or is sorting in general not needed?

**Comment by kayman-mk** [comment]
@viceice I am almost satisfied with the results here. The only problem I have is, that Renovate tries to update the dependency name as well. Let's say the newest version of my lambda is 3 and I have the following file:

```hcl
locals {
  # renovate: datasource=aws-lambda-layer filter={"name": "my-layer", "architecture": "x86_64", "runtime": "python3.7"}
  my_layer_arn = "arn:aws:lambda:us-east-1:123456789012:layer:my-layer:1"
}
```

After Renovate ran on my repository, I see the following change: `  my_layer_arn = "arn:aws:lambda:us-east-3:123456789012:layer:my-layer:3"`. As you can see, the version number was updated (after the last colon) but `us-east-1` was changed into `us-east-3`. Seems that the RegEx manager replaces every `1` by a `3`.

I tried to find some docs explaining the logic here, but didn't found any. Could you shed some light on this, please?

**Comment by JamieMagee** [comment]
What's your renovate config for that upgrade? It looks like it's finding it somewhere in [`auto-replace`](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/branch/auto-replace.ts), but you'd need to debug through to see exactly where.

**Comment by kayman-mk** [comment]
I am developing with VSCode using the devcontainer option. I followed the steps in the [local_development](https://github.com/renovatebot/renovate/blob/main/docs/development/local-development.md) and exported my AWS credentials and the Renovate token. No other special configuration set.

I started renovate with `LOG_LEVEL=debug yarn start kayman-mk/renovate-test` within this feature branch. The only file I added was `.nvmrc` as described in the docs.

**Comment by protoz** [comment]
Not to derail the current design according to the readme using a specific local variable, but it may be good to support the aws provider's data source [aws_lambda_layer_version](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/lambda_layer_version).

``` hcl
data "aws_lambda_layer_version" "datadog_extension" {
  layer_name = "arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Extension"
  version = 37
}
```

By specifying `aws_lambda_layer_version.datadog_extension.arn` it will append the version appropriately as `arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Extension:37` in this case.  That may simplify the regex issues you are having.

**Pull request closed without merging**